File: packages/vault-core/tests/crdt.test.ts
import { describe, expect, it } from "vitest";
import { compareHlc, createClock, decodeHlc, encodeHlc, HLC_ZERO } from "../src/hlc.js";
import { mergeVaults } from "../src/merge.js";
import type { VaultState } from "../src/model.js";
import { emptyVault, fingerprint, itemField, visibleItems } from "../src/model.js";
import type { VaultOp } from "../src/ops.js";
import { applyOp, applyOps } from "../src/ops.js";

const clock = createClock({ node: "test" });
let n = 0;
const id = () => `op-${++n}`;

function put(itemId: string, fields: Record<string, string>, keyringId = "ring"): VaultOp {
  return { kind: "item.put", opId: id(), ts: clock.now(), itemId, keyringId, fields };
}

function build(ops: VaultOp[]): VaultState {
  return applyOps(emptyVault(), ops);
}

describe("hybrid logical clock", () => {
  it("round-trips and sorts lexicographically in causal order", () => {
    const a = encodeHlc({ wall: 1_700_000_000_000, counter: 0, node: "a" });
    const b = encodeHlc({ wall: 1_700_000_000_000, counter: 1, node: "a" });
    const c = encodeHlc({ wall: 1_700_000_000_001, counter: 0, node: "a" });
    expect(decodeHlc(b)).toEqual({ wall: 1_700_000_000_000, counter: 1, node: "a" });
    expect(compareHlc(a, b)).toBe(-1);
    expect(compareHlc(b, c)).toBe(-1);
    expect([c, a, b].sort()).toEqual([a, b, c]);
    expect(HLC_ZERO < a).toBe(true);
  });

  it("advances monotonically even when the wall clock stands still or goes backwards", () => {
    let wall = 1_000;
    const c = createClock({ node: "n", physical: () => wall });
    const first = c.now();
    const second = c.now();
    wall = 500; // clock corrected backwards
    const third = c.now();
    expect(second > first).toBe(true);
    expect(third > second).toBe(true);
  });

  it("sorts a local edit after a remote one it has observed", () => {
    const slow = createClock({ node: "slow", physical: () => 1_000 });
    const fast = createClock({ node: "fast", physical: () => 9_999_999 });
    const remote = fast.now();
    slow.observe(remote);
    expect(slow.now() > remote).toBe(true);
  });

  it("resumes causal time across a restart", () => {
    const before = createClock({ node: "n", physical: () => 5_000 });
    const stamp = before.now();
    // Process restarts and the device clock has been set backwards.
    const after = createClock({ node: "n", physical: () => 100, resume: before.snapshot() });
    expect(after.now() > stamp).toBe(true);
  });
});

describe("merge is a CRDT join", () => {
  const a = build([put("one", { title: "One" }), put("two", { title: "Two" })]);
  const b = build([put("two", { title: "Two edited" }), put("three", { title: "Three" })]);

  it("is idempotent", () => {
    expect(fingerprint(mergeVaults(a, a))).toBe(fingerprint(a));
  });

  it("is commutative", () => {
    expect(fingerprint(mergeVaults(a, b))).toBe(fingerprint(mergeVaults(b, a)));
  });

  it("is associative", () => {
    const c = build([put("four", { title: "Four" })]);
    const left = mergeVaults(mergeVaults(a, b), c);
    const right = mergeVaults(a, mergeVaults(b, c));
    expect(fingerprint(left)).toBe(fingerprint(right));
  });

  it("keeps every item from both sides", () => {
    const merged = mergeVaults(a, b);
    expect(Object.keys(merged.items).sort()).toEqual(["one", "three", "two"]);
  });
});

describe("operations", () => {
  it("resolve per field, so simultaneous edits to one login both survive", () => {
    const base = build([put("chase", { title: "Chase", username: "old", password: "old" })]);
    const withUser = applyOp(base, put("chase", { username: "new-user" }));
    const withPass = applyOp(base, put("chase", { password: "new-pass" }));
    const merged = mergeVaults(withUser, withPass);
    expect(itemField(merged.items.chase!, "username")).toBe("new-user");
    expect(itemField(merged.items.chase!, "password")).toBe("new-pass");
    expect(itemField(merged.items.chase!, "title")).toBe("Chase");
  });

  it("is idempotent when the same operation is replayed", () => {
    const op = put("chase", { password: "abc" });
    const once = applyOp(emptyVault(), op);
    const twice = applyOp(once, op);
    expect(fingerprint(twice)).toBe(fingerprint(once));
  });

  it("lets a later edit beat an earlier delete, and a later delete beat an earlier edit", () => {
    const base = build([put("a", { title: "A" }), put("b", { title: "B" })]);
    const deleteA: VaultOp = { kind: "item.delete", opId: id(), ts: clock.now(), itemId: "a" };
    const afterDelete = applyOp(base, deleteA);
    const revived = applyOp(afterDelete, put("a", { title: "A again" }));
    expect(revived.items.a!.deleted.value).toBe(false);

    const deleteB: VaultOp = { kind: "item.delete", opId: id(), ts: clock.now(), itemId: "b" };
    const gone = applyOp(revived, deleteB);
    expect(gone.items.b!.deleted.value).toBe(true);
  });

  it("hides items whose keyring was deleted without destroying them", () => {
    let state = build([put("x", { title: "X" })]);
    state = applyOp(state, {
      kind: "keyring.put",
      opId: id(),
      ts: clock.now(),
      keyringId: "ring",
      name: "Household",
    });
    expect(visibleItems(state)).toHaveLength(1);
    state = applyOp(state, {
      kind: "keyring.delete",
      opId: id(),
      ts: clock.now(),
      keyringId: "ring",
    });
    expect(visibleItems(state)).toHaveLength(0);
    expect(state.items.x).toBeDefined();
  });

  it("retains a superseded password so an accidental overwrite is recoverable", () => {
    let state = build([put("chase", { password: "original" })]);
    state = applyOp(state, put("chase", { password: "overwritten" }));
    expect(itemField(state.items.chase!, "password")).toBe("overwritten");
    const previous = state.items.chase!.history.find((h) => h.field === "password");
    expect(previous?.value).toBe("original");
  });

  it("caps retained history so a long-lived item cannot grow without bound", () => {
    let state = emptyVault();
    for (let i = 0; i < 40; i += 1) state = applyOp(state, put("chase", { password: `pw-${i}` }));
    expect(state.items.chase!.history.length).toBeLessThanOrEqual(12);
    expect(itemField(state.items.chase!, "password")).toBe("pw-39");
  });
});

describe("fingerprint", () => {
  it("is stable across key insertion order", () => {
    const one = build([put("a", { title: "A" }), put("b", { title: "B" })]);
    const two = mergeVaults(build([put("b", { title: "B" })]), build([put("a", { title: "A" })]));
    // Different construction order, same observable content ordering.
    expect(Object.keys(one.items).sort()).toEqual(Object.keys(two.items).sort());
  });

  it("changes when a register advances", () => {
    const before = build([put("a", { title: "A" })]);
    const after = applyOp(before, put("a", { title: "A2" }));
    expect(fingerprint(after)).not.toBe(fingerprint(before));
  });
});

File: packages/android/vault/src/test/kotlin/app/keyweb/vault/CrdtTest.kt
package app.keyweb.vault

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

private val clock = Clock("test")
private var opCounter = 0
private fun nextId() = "op-${++opCounter}"

private fun put(
    itemId: String,
    fields: Map<ItemField, String>,
    keyringId: String = "ring",
): VaultOp = VaultOp.ItemPut(nextId(), clock.now(), itemId, keyringId, fields)

private fun build(ops: List<VaultOp>): VaultState = applyOps(emptyVault(), ops)

class HlcTest {

    @Test
    fun `round-trips and sorts lexicographically in causal order`() {
        val a = encodeHlc(HlcParts(1_700_000_000_000, 0, "a"))
        val b = encodeHlc(HlcParts(1_700_000_000_000, 1, "a"))
        val c = encodeHlc(HlcParts(1_700_000_000_001, 0, "a"))
        assertEquals(HlcParts(1_700_000_000_000, 1, "a"), decodeHlc(b))
        assertTrue(a < b)
        assertTrue(b < c)
        assertEquals(listOf(a, b, c), listOf(c, a, b).sorted())
        assertTrue(HLC_ZERO < a)
    }

    @Test
    fun `advances monotonically even when the wall clock stalls or goes backwards`() {
        var wall = 1_000L
        val c = Clock("n", { wall })
        val first = c.now()
        val second = c.now()
        wall = 500 // clock corrected backwards
        val third = c.now()
        assertTrue(second > first)
        assertTrue(third > second)
    }

    @Test
    fun `sorts a local edit after a remote one it has observed`() {
        val slow = Clock("slow", { 1_000 })
        val fast = Clock("fast", { 9_999_999 })
        val remote = fast.now()
        slow.observe(remote)
        assertTrue(slow.now() > remote)
    }

    @Test
    fun `resumes causal time across a restart`() {
        val before = Clock("n", { 5_000 })
        val stamp = before.now()
        // Process restarts and the device clock has been set backwards.
        val after = Clock("n", { 100 }, resume = before.snapshot())
        assertTrue(after.now() > stamp)
    }
}

class MergeLawsTest {
    private val a = build(listOf(put("one", mapOf(Fields.TITLE to "One")), put("two", mapOf(Fields.TITLE to "Two"))))
    private val b = build(listOf(put("two", mapOf(Fields.TITLE to "Two edited")), put("three", mapOf(Fields.TITLE to "Three"))))

    @Test
    fun `is idempotent`() {
        assertEquals(fingerprint(a), fingerprint(mergeVaults(a, a)))
    }

    @Test
    fun `is commutative`() {
        assertEquals(fingerprint(mergeVaults(a, b)), fingerprint(mergeVaults(b, a)))
    }

    @Test
    fun `is associative`() {
        val c = build(listOf(put("four", mapOf(Fields.TITLE to "Four"))))
        assertEquals(
            fingerprint(mergeVaults(mergeVaults(a, b), c)),
            fingerprint(mergeVaults(a, mergeVaults(b, c))),
        )
    }

    @Test
    fun `keeps every item from both sides`() {
        assertEquals(listOf("one", "three", "two"), mergeVaults(a, b).items.keys.sorted())
    }
}

class OpsTest {

    @Test
    fun `resolves per field, so simultaneous edits to one login both survive`() {
        val base = build(
            listOf(
                put(
                    "chase",
                    mapOf(
                        Fields.TITLE to "Chase",
                        Fields.USERNAME to "old",
                        Fields.PASSWORD to "old",
                    ),
                ),
            ),
        )
        val withUser = applyOp(base, put("chase", mapOf(Fields.USERNAME to "new-user")))
        val withPass = applyOp(base, put("chase", mapOf(Fields.PASSWORD to "new-pass")))
        val merged = mergeVaults(withUser, withPass)
        val item = assertNotNull(merged.items["chase"])
        assertEquals("new-user", item.field(Fields.USERNAME))
        assertEquals("new-pass", item.field(Fields.PASSWORD))
        assertEquals("Chase", item.field(Fields.TITLE))
    }

    @Test
    fun `is idempotent when the same operation is replayed`() {
        val op = put("chase", mapOf(Fields.PASSWORD to "abc"))
        val once = applyOp(emptyVault(), op)
        val twice = applyOp(once, op)
        assertEquals(fingerprint(once), fingerprint(twice))
    }

    @Test
    fun `a later edit beats an earlier delete, and a later delete beats an earlier edit`() {
        val base = build(listOf(put("a", mapOf(Fields.TITLE to "A")), put("b", mapOf(Fields.TITLE to "B"))))
        val afterDelete = applyOp(base, VaultOp.ItemDelete(nextId(), clock.now(), "a"))
        val revived = applyOp(afterDelete, put("a", mapOf(Fields.TITLE to "A again")))
        assertEquals(false, revived.items["a"]?.deleted?.value)

        val gone = applyOp(revived, VaultOp.ItemDelete(nextId(), clock.now(), "b"))
        assertEquals(true, gone.items["b"]?.deleted?.value)
    }

    @Test
    fun `hides items whose keyring was deleted without destroying them`() {
        var state = build(listOf(put("x", mapOf(Fields.TITLE to "X"))))
        state = applyOp(state, VaultOp.KeyringPut(nextId(), clock.now(), "ring", "Household"))
        assertEquals(1, visibleItems(state).size)
        state = applyOp(state, VaultOp.KeyringDelete(nextId(), clock.now(), "ring"))
        assertEquals(0, visibleItems(state).size)
        assertNotNull(state.items["x"])
    }

    @Test
    fun `retains a superseded password so an accidental overwrite is recoverable`() {
        var state = build(listOf(put("chase", mapOf(Fields.PASSWORD to "original"))))
        state = applyOp(state, put("chase", mapOf(Fields.PASSWORD to "overwritten")))
        assertEquals("overwritten", state.items["chase"]?.field(Fields.PASSWORD))
        val previous = state.items["chase"]?.history?.firstOrNull { it.field == Fields.PASSWORD }
        assertEquals("original", previous?.value)
    }

    @Test
    fun `caps retained history so a long-lived item cannot grow without bound`() {
        var state = emptyVault()
        repeat(40) { i -> state = applyOp(state, put("chase", mapOf(Fields.PASSWORD to "pw-$i"))) }
        val item = assertNotNull(state.items["chase"])
        assertTrue(item.history.size <= HISTORY_LIMIT)
        assertEquals("pw-39", item.field(Fields.PASSWORD))
    }
}

class FingerprintTest {

    @Test
    fun `changes when a register advances`() {
        val before = build(listOf(put("a", mapOf(Fields.TITLE to "A"))))
        val after = applyOp(before, put("a", mapOf(Fields.TITLE to "A2")))
        assertNotEquals(fingerprint(before), fingerprint(after))
    }

    @Test
    fun `does not depend on insertion order`() {
        val one = build(listOf(put("a", mapOf(Fields.TITLE to "A"))))
        val two = build(listOf(put("b", mapOf(Fields.TITLE to "B"))))
        assertEquals(fingerprint(mergeVaults(one, two)), fingerprint(mergeVaults(two, one)))
    }
}
